Improvements to groups UI.

m-e-l-u-h-a-n opened the pull request:
Implments changes suggested in #24443. Following changes are completed:

- [x] Tooltip on `+`.
- [x] Something is wrong with the Actions column in the Members tab.
- [x] Add a "Delete group" tooltip to the trash can button.
- [x] Alphabetize groups by name (ignoring capitalization) in the left sidebar.
- [x] Add "Your groups" / "All groups" tabs in the left panel, equivalent to "Subscribed" / "All streams" (This has some follow-ups that are being addressed.) 

Fixes: Partially addresses #24443.

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
@m-e-l-u-h-a-n cool! Let us know when this is ready for review.

m-e-l-u-h-a-n commented:
@timabbott this PR is now complete and ready for review. All the points mentioned in #24443 are completed except for:
> Add a Join/Leave button to the left of the trash can. (Tooltips: "Join group" / "Leave group")

As this would require API changes about who is allowed to join/leave the group. I would start the discussion regarding this along with other permissions tab related API changes on CZO.
Apart from that I have included a lot of minor fixes, refactoring and performance update related commits to this PR so that after merging this PR we would be in a state to try out this UI on CZO.

timabbott commented on web/src/user_groups_settings_ui.js:
```diff
@@ -54,13 +54,13 @@ export const show_user_group_settings_pane = {
     },
     create_user_group() {
-        $(".nothing-selected, .settings, #user-group-creation").hide();
+        $(".nothing-selected, .settings").hide();
```
Both `.settings` and the `.nothing-selected` selector seem badly scoped -- we want to just interact with the user groups variant of these, not other elements...,

m-e-l-u-h-a-n replied:
Fixed this to only query these classes within `#groups_overlay`.

timabbott commented:
> As this would require API changes about who is allowed to join/leave the group. I would start the discussion regarding this along with other permissions tab related API changes on CZO.

OK, that makes sense -- can you open an issue for just that remaining task? 

@sahil839 I did a read through this and only noticed one thing, but can you also do a review? My hope is to merge this pretty quickly so we can keep iterating on this part of the project starting from a much less buggy initial state.

sahil839 reviewed (commented):
Posted some comments.

timabbott commented:
@m-e-l-u-h-a-n what's the status of this PR?

m-e-l-u-h-a-n commented:
Ohh sorry for the delay on this, I've my end-term exams going this week.
I've made fixes for some of the review suggestions and would push the updates soon after this week as exams end.

timabbott commented:
@m-e-l-u-h-a-n is this ready for a next review? As a general rule, it's always good to post a detailed comment on the changes; https://zulip.readthedocs.io/en/latest/contributing/reviewable-prs.html has some documentation that you may find helpful; you may also find it useful to look at https://github.com/zulip/zulip/pulls/laurynmm for some good examples.

timabbott commented on web/src/tippyjs.js:
```diff
@@ -470,4 +470,13 @@ export function initialize() {
+            defaultMessage: "Create new user group",
+        }),
+        placement: "bottom",
+        appendTo: () => document.body,
```
Having now seen the commit message, please put a brief version of the explanation in a comment here, since otherwise I bet someone will "simplify" it back to the broken way :)

m-e-l-u-h-a-n replied:
Added a comment to explain these changes.

m-e-l-u-h-a-n commented on web/templates/user_group_settings/user_group_settings.hbs:
```diff
@@ -15,7 +15,7 @@
                 </div>
                 <div class="group_change_property_info alert-notification"></div>
-                <div class="button-group" {{#unless can_edit}}style="display:none"{{/unless}}>
+                <div class="button-group {{#unless can_edit}} hide {{/unless}}">
```
Yeah sorry just calling `show()` and and `hide()` on correct element did the work. I've removed this change as it was no longer needed.

m-e-l-u-h-a-n commented on web/src/user_group_edit.js:
```diff
@@ -191,10 +191,14 @@ export function is_editing_group(desired_group_id) {
+    if (is_editing_group(group_id)) {
+        open_right_panel_empty();
+    }
+    user_group_settings_ui.redraw_user_group_list();
```
We need to update the list of items maintained by `ListWidget` along with deleting the row from DOM. If we simply delete the row from DOM and not update the group list data maintained by `ListWidget` in next re-render would throw errors as the user group is deleted but still exist in `ListWidget`. So call to `user_group_settings_ui.redraw_user_group_list()` updates the group list data maintained by `ListWidget` according to active tab (Your groups/All groups ) and also updates DOM according to the list.

sahil839 replied:
Sounds good.

m-e-l-u-h-a-n commented on web/src/user_group_edit.js:
```diff
@@ -114,6 +114,24 @@ export function handle_member_edit_event(group_id) {
     user_group_edit_members.update_member_list_widget(group_id, members);
 
+    // update display of group-row.
+    user_group_settings_ui.redraw_user_group_list();
```
`ListWidget` also maintains the list of items and that needs to be updated. If we not update the group list maintained by the `ListWidget` it would remove the newly subscribed group or would again show some group for which the user was unsubscribed during some re-render due to interaction with search bar.
So Listwidget when we updates data maintained by `ListWidget` it automatically updates the list of items shown on DOM based on updated data.

I've made changes to optimize this and make calls to `user_group_settings_ui.redraw_user_group_list()` only when needed and also added a comment to explain the changes.

m-e-l-u-h-a-n commented on web/templates/user_group_settings/user_group_settings_overlay.hbs:
```diff
@@ -9,19 +9,21 @@
-                    <button type="button" class="btn clear_search_button" id="clear_search_group_name">
-                        <i class="fa fa-remove" aria-hidden="true"></i>
-                    </button>
+                <div class="search-container">
```
I've named it so that it is similar to stream settings. Should we update this for both #groups and #streams overlay.

sahil839 replied:
We should update it for both. But can also be done as a follow-up, this PR is already large.

m-e-l-u-h-a-n commented on web/src/user_group_edit.js:
```diff
@@ -163,10 +163,24 @@ export function setup_group_settings(node) {
+    } else if (tab_key_value === "your-groups") {
+        browser_history.update("#groups/your");
+    } else {
+        blueslip.debug(`Unknown tab_key_value: ${tab_key_value} for groups overlay.`);
```
Made the changes to not update the URL if a group is opened in the right panel.

m-e-l-u-h-a-n commented:
> @m-e-l-u-h-a-n is this ready for a next review? As a general rule, it's always good to post a detailed comment on the changes; https://zulip.readthedocs.io/en/latest/contributing/reviewable-prs.html has some documentation that you may find helpful; you may also find it useful to look at https://github.com/zulip/zulip/pulls/laurynmm for some good examples.

I was waiting for checks to pass, before addressing review comments but as it got bit late (4:00 AM IST), I went to sleep :smile: and had put the task about answering review comments in the morning.

This is ready for another review now.

alya commented:
Thanks! I noticed that the checkmarks to subscribe/unsubscribe don't work, but if that was already the case before this PR, we can add it to the issue instead of solving it here. I don't know if we want to finish up #22252 before fixing this.

Looks good to me in manual testing otherwise.

m-e-l-u-h-a-n commented:
> the checkmarks to subscribe/unsubscribe don't work.

It was same before this PR. I've opened https://github.com/zulip/zulip/issues/25538 to track this, and is the next thing I am planning to work on.

alya commented:
OK, great. @sahil839 do you have any more feedback on this one, or should @timabbott take a look next?

sahil839 commented on web/src/user_groups_settings_ui.js:
```diff
@@ -129,7 +129,6 @@ export function add_group_to_table(group) {
         can_edit: user_group_edit.can_edit(group.id),
     });
 
-
```
This line was added in the previous commit itself. We should remove it there only to avoid this change in this commit.

m-e-l-u-h-a-n replied:
Ohh, fixed this.

sahil839 commented on web/src/user_group_edit.js:
```diff
@@ -114,6 +114,31 @@ export function handle_member_edit_event(group_id) {
+            $sub_unsub_button.removeClass("checked");
+            $sub_unsub_button.addClass("disabled");
+        }
+    }
```
This can probably be moved to `user_group_ui_updates`. Should be fine to do as a follow-up as well. General pattern that I have noticed is - `stream_settings_ui` contains functions which are called on recieving events and then those function calls the appropriate function in `stream_data` to update the data objects and function in `stream_ui_updates` to update the UI.

This pattern is still not followed strictly for stream settings, but we can try doing so here as it sounds like a nice pattern to follow which will help in finding the code easily. We might also make sure we avoid circular imports as we would be converting these modules to `Typescript`.

m-e-l-u-h-a-n replied:
I think user_group modules need significant refactoring as there are lot of cyclic imports and many functions do not align with the idea (that sounds really good) you mentioned, that deserve it's own PR. 
Please comment in case I should complete the refactors and add the commits to this PR only.

sahil839 reviewed (commented):
Posted one minor comment and one for code structure changes that could be done as a follow-up. Looks good to me otherwise.

m-e-l-u-h-a-n commented:
I've addressed the all the review comments. There are a few follow-ups that appear from the review discussion on this PR and its issue thread. Summarizing them in this comment, please let me any of those need to be added to this PR before merging this:

- [ ] Making checkmarks on group rows clickable so that users can easily join or leave the group. https://github.com/zulip/zulip/issues/25538 tracks this.
- [ ] Minor improvement in readability of `user_group_setttings_overlay` handlebar. Discussed here: https://github.com/zulip/zulip/pull/24766#discussion_r1191969035
- [ ] Refactoring `user_group_settings` module to avoid cyclic imports and follow a predictable pattern mentioned here: https://github.com/zulip/zulip/pull/24766#discussion_r1192002893

This is ready for another review.

timabbott commented:
I resolved the merge conflicts and one new linter error revealed by rebasing; with those changes, this lgtm, marked to merge once CI passes. Thanks for all the work on this @m-e-l-u-h-a-n and sorry I missed integrating this during my leave this summer.

@sahil839 can you make sure everything in https://github.com/zulip/zulip/pull/24766#issuecomment-1546053872 gets turned into issues in the "user groups" area label (https://github.com/zulip/zulip/labels/area%3A%20settings%20%28user%20groups%29) so we don't lose track of them?

It also looks like #24443 has a few items still not checked off; I think we should aim to do a quick PR to make the wording changes and then close the issue to simplify the issue tracker aroundn this topic.

sahil839 commented:
I opened #26779 for the refactoring follow-ups, would be good if you could have a look at it once. And one of the follow-ups is already tracked by #25538.

I have opened #26782 for one remaining follow-up mentioned in that comment and to fix remaining parts of #24443.
